From 8080ce176dce38c5b2ca41a0e6485a6d901b2b3e Mon Sep 17 00:00:00 2001
From: Aaradhana Sahu <quic_aarasahu@quicinc.com>
Date: Fri, 3 Mar 2023 15:15:03 +0530
Subject: [PATCH] ath12k: fix firmware crash issue in raw mode with software
 encryption

Currently, we are not setting "WMI_UNIFIED_VDEV_START_HW_ENCRYPTION_DISABLED"
flag during vdev start. this will trigger firmware assert because in SW
encryption encrypt/decrypt should be done by software.

So, Fix this issue by setting "WMI_UNIFIED_VDEV_START_HW_ENCRYPTION_DISABLED"
flag to avoid hardware encryption/decryption.

Signed-off-by: Aaradhana Sahu <quic_aarasahu@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/wmi.c | 4 ++++
 drivers/net/wireless/ath/ath12k/wmi.h | 1 +
 2 files changed, 5 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/wmi.c b/drivers/net/wireless/ath/ath12k/wmi.c
index f57d926..f39066f 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -1148,6 +1148,7 @@ int ath12k_wmi_vdev_start(struct ath12k *ar, struct wmi_vdev_start_req_arg *arg,
 	void *ptr;
 	int ret, len, ml_arg_size = 0;
 	u8 i;
+	struct ath12k_hw_group *ag = ar->ab->ag;
 
 	if (WARN_ON(arg->ssid_len > sizeof(cmd->ssid.ssid)))
 		return -EINVAL;
@@ -1192,6 +1193,9 @@ int ath12k_wmi_vdev_start(struct ath12k *ar, struct wmi_vdev_start_req_arg *arg,
 
 	cmd->flags |= cpu_to_le32(WMI_VDEV_START_LDPC_RX_ENABLED);
 
+	if (test_bit(ATH12K_FLAG_HW_CRYPTO_DISABLED, &ag->dev_flags))
+		cmd->flags |= cpu_to_le32(WMI_VDEV_START_HW_ENCRYPTION_DISABLED);
+
 	ptr = skb->data + sizeof(*cmd);
 	chan = ptr;
 
diff --git a/drivers/net/wireless/ath/ath12k/wmi.h b/drivers/net/wireless/ath/ath12k/wmi.h
index 9257720..3fa2a7c 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.h
+++ b/drivers/net/wireless/ath/ath12k/wmi.h
@@ -2991,6 +2991,7 @@ struct wmi_vdev_down_cmd {
 #define WMI_VDEV_START_HIDDEN_SSID  BIT(0)
 #define WMI_VDEV_START_PMF_ENABLED  BIT(1)
 #define WMI_VDEV_START_LDPC_RX_ENABLED BIT(3)
+#define WMI_VDEV_START_HW_ENCRYPTION_DISABLED BIT(4)
 
 struct wmi_ssid {
 	u32 ssid_len;
-- 
2.17.1

